spring-ai-core/src/main/java/org/springframework/ai/chat/observation/ChatModelObservationSupport.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2023-2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.ai.chat.observation;
+
+import java.util.Optional;
+import java.util.function.Consumer;
+
+import io.micrometer.observation.Observation;
+
+import org.springframework.ai.chat.model.ChatModel;
+import org.springframework.ai.chat.model.ChatResponse;
+import org.springframework.lang.Nullable;
+
+/**
+ * Support class for processing {@link ChatModel} Micrometer {@link Observation
+ * Observations}.
+ *
+ * @author John Blum
+ * @see ChatModel
+ * @see Observation
+ * @since 1.0.0
+ */
+public abstract class ChatModelObservationSupport {
+
+	public static Optional<ChatModelObservationContext> getObservationContext(@Nullable Observation observation) {
+
+		// Avoid unnecessary construction of an Optional if Observations are not enabled
+		// (aka NOOP).
+		return Observation.NOOP.equals(observation) ? Optional.empty()
```

Comment:
Honestly, I went back and forth on this.

I realize that in the context of Micrometer, the `Observation` is likely to never to be `null`. But, arguably, contracts can change (or break). So, over many years of systems development, I learned to code defensively... assume the worse, fail gracefully and deliberately, not unexpectedly.

GENERAL THINKING:

[This](https://github.com/jxblum/spring-ai/blob/pr/observation/spring-ai-core/src/main/java/org/springframework/ai/chat/observation/ChatModelObservationSupport.java#L43-L44) will never fail (providing the `equals(..)` method is coded correctly), where as, `object.someMethod(..)` can throw an NPE at runtime if our library is not used properly, even despite the "nullability" contract. Additionally, the "nullability" contract is not always apparent (despite implied or explicit annotations, which rely on tooling to do the right thing, e.g. warn the user). The support class does not make any assumptions about how the `Observation` was created (e.g. with a `ObservationDocumentation`, or otherwise, or even at all). The point is, we don't know how our method will be called.

This same principle applies to methods that return arrays or collections. It is a recommended practice for classes to never return a `null` array or collection, but rather an empty one. Still, I never take that for granted and assume a 3rd party library (or even my own code; mistakes happen) will always do the right thing, so I guard against the possibility of `null` arrays/collections.

For example:

```
List<User> users = userService.findByName("...");
nullSafeList(users).forEach(user -> ...);
```

Where:

```
static <T> List<T> nullSafeList(List<T> list) {
  return list != null ? list : Collections.emptyList();
}
```

Rather than:

```
List<User> users = userService.findByName("...");
users.forEache(user -> ...);
```

The later is bad if the contract of `UserService.findByName(..):List<User>` is broken, returning `null` instead of empty.

This might seem like paranoia, until it isn't.

Safety first.